sdk/core/System.ClientModel/src/Pipeline/PipelineTransport.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading.Tasks;
+
+namespace System.ClientModel.Primitives;
+
+// TODO: MessageClassifier will come in a later PR.
+
+public abstract class PipelineTransport : PipelinePolicy
+{
+    /// <summary>
+    /// TBD: needed for inheritdoc.
+    /// </summary>
+    /// <param name="message"></param>
+    public void Process(PipelineMessage message)
+    {
+        ProcessCore(message);
+
```

Comment:
Can/should buffering be done here as opposed to doing it in a policy?

Reply:
Interesting.  I think I considered this at some point and decided it wasn't possible, but we've iterated on the design here, and it might be possible now since we've moved to the template pattern.  Let me look again and get back to you if I find a blocker.

Reply:
I put up a quick prototype of how we might achieve this here: https://github.com/Azure/azure-sdk-for-net/pull/41149

Overall, I like the idea a lot, so let's consider the open questions/risks:

For context, the Azure.Core response buffering policy does essentially three things:
  1. Buffers the response if the client hasn't said not to
  2. If it _doesn't_ buffer it, it wraps it in a ReadTimeoutStream
  3. Implements the network timeout for the whole pipeline

If the CI's on that PR go green, then it should be technically possible to move 1 & 3 into the transport.  I didn't move 2 into the feature branch I think because it brought a lot of dependencies and we were going fast for the demo, but if we decide to do this, I'll include that in this PR.

There are two wrinkles I see here offhand:
1. We'll need to find a place for the client-options network timeout value to live so each client can have its own -- I don't think it can live on the transport since store a shared instance of that.  I don't have a deep understanding of how we share transports across clients, so if we can find a convenient place to store this that's not the transport, that would be the simplest approach.  If we're happy with creating a new instance of a transport per pipeline based on client options, that would work too, and then that value would rightly live with the timeout logic.
2. In the Azure.Core default pipeline, the response buffering policy lives before the RequestActivityPolicy (distributed tracing) and the client/user-specified BeforeTransport policies.  Offhand, I don't see any reason that those policies would rely on the response _not_ being buffered, but if we think there's a risk of breaking changes from moving the functional pieces described above to _after_ these policies, we would want to consider those as well.

Reply:
@JoshLove-msft, we've been thinking we'd change that static method to `DisableBuffering(message)` - I think that could move to the transport - we didn't want to put it on the client options-equivalent since it's such an edge case.

Reply:
Ok, one of the test failures is because we have [Azure.Core tests that validate that reading from the stream set by the transport (a network stream) gives an IOException](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/tests/TransportFunctionalTests.cs#L974).  If we move the network timeout functionality into the transport, this gets wrapped in a TaskCanceledException.

We would need to decide whether this is an acceptable behavioral change.  HttpClientTransport is a public type in Azure.Core, so we would be throwing a new exception type in this case, so would be a breaking change in Azure.Core 2.0.

We could look at moving the network timeout functionality into a separate policy (which would replace the response buffering policy), but this investigation surfaces the risk that making behavioral changes to a GA type could be problematic, i.e. if any end-user code relies on HttpClientTransport.Process leaving the message's response content stream unbuffered.  Our buffering logic would be calling dispose on that stream.  We could advise users on how to disable buffering if they wanted that behavior, but it would be a breaking change.

What do you think, @KrzysztofCwalina?

Reply:
Filing https://github.com/Azure/azure-sdk-for-net/issues/41198 to track this work